// syncano/lib/models/dataobject.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _difference2 = require('lodash/difference');

var _difference3 = _interopRequireDefault(_difference2);

var _union2 = require('lodash/union');

var _union3 = _interopRequireDefault(_union2);

var _concat2 = require('lodash/concat');

var _concat3 = _interopRequireDefault(_concat2);

var _isArray2 = require('lodash/isArray');

var _isArray3 = _interopRequireDefault(_isArray2);

var _add2 = require('lodash/add');

var _add3 = _interopRequireDefault(_add2);

var _isNumber2 = require('lodash/isNumber');

var _isNumber3 = _interopRequireDefault(_isNumber2);

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _keys2 = require('lodash/keys');

var _keys3 = _interopRequireDefault(_keys2);

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _querySet = require('../querySet');

var _querySet2 = _interopRequireDefault(_querySet);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataObjectQuerySet = (0, _stampit2.default)().compose(_querySet2.default).methods({
  /**
    * Filters DataObjects.
     * @memberOf QuerySet
    * @instance
     * @param {Object} filters
    * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).filter({ field_name: { _contains: 'Lord Of The Rings' }}).then(function(dataobjects) {});
     */

  filter: function filter() {
    var filters = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    this.query['query'] = JSON.stringify(filters);
    return this;
  },

  /**
    * Orders DataObject by field.
     * @memberOf QuerySet
    * @instance
     * @param {String} field
    * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).orderBy('author').then(function(dataobjects) {});
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).orderBy('-author').then(function(dataobjects) {});
     */
  orderBy: function orderBy(field) {
    this.query['order_by'] = field;
    return this;
  },

  /**
  * Adds an array to an array field.
   * @memberOf QuerySet
  * @instance
   * @param {Object} properties lookup properties used for path resolving
  * @param {Object} field to add to.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * DataObject.please().add({instanceName: 'my-instance', className: 'my-class', id: 1}, {array_field: [1,2]})
   */
  add: function add() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var object = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var payload = {};
    payload[(0, _keys3.default)(object)[0]] = { _add: object[(0, _keys3.default)(object)[0]] };
    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.payload = JSON.stringify(payload);

    this.method = 'PATCH';
    this.endpoint = 'detail';
    return this;
  },

  /**
  * Adds an array to an array field without duplicate values.
   * @memberOf QuerySet
  * @instance
   * @param {Object} properties lookup properties used for path resolving
  * @param {Object} field to add to.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * DataObject.please().add({instanceName: 'my-instance', className: 'my-class', id: 1}, {array_field: [1,2]})
   */
  addUnique: function addUnique() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var object = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var payload = {};
    payload[(0, _keys3.default)(object)[0]] = { _addunique: object[(0, _keys3.default)(object)[0]] };
    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.payload = JSON.stringify(payload);

    this.method = 'PATCH';
    this.endpoint = 'detail';
    return this;
  },

  /**
    * Filters DataObjects using _is.
     * @memberOf QuerySet
    * @instance
     * @param {String} field name
    * @param {Object} filters
    * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({instanceName: 'my-instance', className: 'books'}).is('authors', { name: { _eq: 'Stephen King'}})
     */
  is: function is(field) {
    var object = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var query = {};
    query[field] = { _is: object };
    return this.filter(query);
  },

  /**
    * Filters DataObjects using _contains.
     * @memberOf QuerySet
    * @instance
     * @param {String} field name
    * @param {Array} array of ids
    * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({instanceName: 'my-instance', className: 'books'}).contains('authors', [1, 2, 3])
     */
  contains: function contains(field) {
    var array = arguments.length <= 1 || arguments[1] === undefined ? [] : arguments[1];

    var query = {};
    query[field] = { _contains: array };
    return this.filter(query);
  },

  /**
  * Subtracts an array from an array field.
   * @memberOf QuerySet
  * @instance
   * @param {Object} properties lookup properties used for path resolving
  * @param {Object} field to subtract from.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * DataObject.please().remove({instanceName: 'my-instance', className: 'my-class', id: 1}, {array_field: [1,2]})
   */
  remove: function remove() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var object = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var payload = {};
    payload[(0, _keys3.default)(object)[0]] = { _remove: object[(0, _keys3.default)(object)[0]] };
    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.payload = JSON.stringify(payload);

    this.method = 'PATCH';
    this.endpoint = 'detail';
    return this;
  },

  /**
  * Increments single object based on provided arguments
   * @memberOf QuerySet
  * @instance
   * @param {Object} properties lookup properties used for path resolving
  * @param {Object} field to increment.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * DataObject.please().increment({instanceName: 'my-instance', className: 'my-class', id: 1}, {views: 1})
   */
  increment: function increment() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var object = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    var payload = {};
    payload[(0, _keys3.default)(object)[0]] = { _increment: object[(0, _keys3.default)(object)[0]] };
    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.payload = JSON.stringify(payload);

    this.method = 'PATCH';
    this.endpoint = 'detail';
    return this;
  },

  /**
    * Filters dataobjects by a geopoint field.
     * @memberOf QuerySet
    * @instance
     * @param {Object} coordinates
    * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).near({ geopoint_field_name: { latitude: POINT_LATITUDE, longitude: POINT_LONGITUDE }}).then(function(dataobjects) {});
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).near({ geopoint_field_name: { latitude: POINT_LATITUDE, longitude: POINT_LONGITUDE, distance_in_kilometers: DISTANCE_IN_KILOMETERS }}).then(function(dataobjects) {});
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).near({ geopoint_field_name: { latitude: POINT_LATITUDE, longitude: POINT_LONGITUDE, distance_in_miles: DISTANCE_IN_MILES }}).then(function(dataobjects) {});
     */
  near: function near() {
    var object = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    var query = {};
    query[(0, _keys3.default)(object)[0]] = { _near: object[(0, _keys3.default)(object)[0]] };
    return this.filter(query);
  },

  /**
    * Returns DataObject count.
     * @memberOf QuerySet
    * @instance
     * @returns {QuerySet}
     * @example {@lang javascript}
    * DataObject.please().list({ instanceName: 'test-instace', className: 'test-class' }).count().then(function(response) {});
     */
  count: function count() {
    this.query['include_count'] = true;
    this.raw();
    return this;
  }
});

var DataObjectMeta = (0, _base.Meta)({
  name: 'dataobject',
  pluralName: 'dataobjects',
  endpoints: {
    'detail': {
      'methods': ['delete', 'patch', 'post', 'get'],
      'path': '/v1.1/instances/{instanceName}/classes/{className}/objects/{id}/'
    },
    'list': {
      'methods': ['post', 'get'],
      'path': '/v1.1/instances/{instanceName}/classes/{className}/objects/'
    }
  }
});

var DataobjectConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    }
  },
  className: {
    presence: true,
    string: true
  },
  owner: {
    numericality: true
  },
  owner_permissions: {
    inclusion: ['none', 'read', 'write', 'full']
  },
  group: {
    numericality: true
  },
  group_permissions: {
    inclusion: ['none', 'read', 'write', 'full']
  },
  other_permissions: {
    inclusion: ['none', 'read', 'write', 'full']
  },
  channel: {
    string: true
  },
  channel_room: {
    string: true
  }
};

/**
 * OO wrapper around instance data objects {@link http://docs.syncano.com/v4.0/docs/view-data-objects endpoint}.
 * This model is special because each instance will be **dynamically populated** with fields defined in related {@link Class} schema attribute.
 * @constructor
 * @type {DataObject}

 * @property {Number} id
 * @property {String} instanceName
 * @property {String} className
 * @property {Number} revision
 * @property {Number} [owner = null]
 * @property {String} [owner_permissions = null]
 * @property {Number} [group = null]
 * @property {String} [group_permissions = null]
 * @property {String} [other_permissions = null]
 * @property {String} [channel = null]
 * @property {String} [channel_room = null]

 * @property {String} [description = null]
 * @property {String} [links = {}]
 * @property {Date} [created_at = null]
 * @property {Date} [updated_at = null]
 */
var DataObject = (0, _stampit2.default)().compose(_base.Model).setMeta(DataObjectMeta).methods({
  /**
  * Increments single object field based on provided arguments
   * @memberOf QuerySet
  * @instance
   * @param {String} field name.
  * @param {Number} value to increment the field by,
  * @returns {QuerySet}
   * @example {@lang javascript}
  * Object.increment('views', 1);
   */

  increment: function increment(field, by) {
    if (!(0, _isNumber3.default)(this[field])) return _bluebird2.default.reject(new Error('The ' + field + ' is not numeric.'));
    if (!(0, _isNumber3.default)(by)) return _bluebird2.default.reject(new Error('The provided value is not numeric.'));

    this[field] += (0, _add3.default)(this[field], by);

    return this.save();
  },

  /**
  * Adds an array to an array field.
   * @memberOf QuerySet
  * @instance
   * @param {String} field name.
  * @param {Array} array to add to the field.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * Object.add('authors', [1,2,3]);
   */
  add: function add(field, array) {
    if (!(0, _isArray3.default)(this[field].value)) return _bluebird2.default.reject(new Error('The ' + field + ' is not an array.'));
    if (!(0, _isArray3.default)(array)) return _bluebird2.default.reject(new Error('The provided value is not an array.'));

    this[field] = (0, _concat3.default)(this[field].value, array);

    return this.save();
  },

  /**
  * Adds an array to an array field without duplicate values.
   * @memberOf QuerySet
  * @instance
   * @param {String} field name.
  * @param {Array} array to add to the field.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * Object.add('authors', [1,2,3]);
   */
  addUnique: function addUnique(field, array) {
    if (!(0, _isArray3.default)(this[field].value)) return _bluebird2.default.reject(new Error('The ' + field + ' is not an array.'));
    if (!(0, _isArray3.default)(array)) return _bluebird2.default.reject(new Error('The provided value is not an array.'));

    this[field] = (0, _union3.default)(this[field].value, array);

    return this.save();
  },

  /**
  * Subtracts an array from an array field.
   * @memberOf QuerySet
  * @instance
   * @param {String} field name.
  * @param {Array} array to subtract from the field.
  * @returns {QuerySet}
   * @example {@lang javascript}
  * Object.remove('authors', [1]);
   */
  remove: function remove(field, array) {
    if (!(0, _isArray3.default)(this[field].value)) return _bluebird2.default.reject(new Error('The ' + field + ' is not an array.'));
    if (!(0, _isArray3.default)(array)) return _bluebird2.default.reject(new Error('The provided value is not an array.'));

    this[field] = (0, _difference3.default)(this[field].value, array);

    return this.save();
  }
}).setQuerySet(DataObjectQuerySet).setConstraints(DataobjectConstraints);

exports.default = DataObject;
module.exports = exports['default'];
