// syncano/lib/models/plan.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var PlanQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.Get, _querySet.List).methods({
  subscribe: function subscribe() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var commitment = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.method = 'POST';
    this.payload = { commitment: commitment };
    this.endpoint = 'subscribe';

    return this;
  }
});

var PlanMeta = (0, _base.Meta)({
  name: 'plan',
  pluralName: 'plans',
  endpoints: {
    'detail': {
      'methods': ['get'],
      'path': '/v1.1/billing/plans/{name}/'
    },
    'list': {
      'methods': ['get'],
      'path': '/v1.1/billing/plans/'
    },
    'subscribe': {
      'methods': ['post'],
      'path': '/v1.1/billing/plans/{name}/subscribe/'
    }
  }
});
/**
 * OO wrapper around Plan.
 * @constructor
 * @type {Subscription}

 * @property {String} name
 * @property {Object} pricing
 * @property {Object} options
 */
var Plan = (0, _stampit2.default)().compose(_base.Model).setQuerySet(PlanQuerySet).setMeta(PlanMeta).methods({
  subscribe: function subscribe() {
    var commitment = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    var meta = this.getMeta();
    var path = meta.resolveEndpointPath('subscribe', this);

    return this.makeRequest('POST', path, { commitment: commitment });
  }
});

exports.default = Plan;
module.exports = exports['default'];
