// syncano/lib/models/restore.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign2 = require('lodash/assign');

var _assign3 = _interopRequireDefault(_assign2);

var _stampit = require('stampit');

var _stampit2 = _interopRequireDefault(_stampit);

var _base = require('./base');

var _querySet = require('../querySet');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RestoreQuerySet = (0, _stampit2.default)().compose(_querySet.BaseQuerySet, _querySet.List, _querySet.Get).methods({
  restore: function restore() {
    var properties = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var payload = arguments.length <= 1 || arguments[1] === undefined ? {} : arguments[1];

    this.properties = (0, _assign3.default)({}, this.properties, properties);
    this.method = 'POST';
    this.payload = payload;
    this.endpoint = 'restore';
    return this;
  }
});

var RestoreMeta = (0, _base.Meta)({
  name: 'restore',
  pluralName: 'restores',
  endpoints: {
    'detail': {
      'methods': ['get'],
      'path': '/v1.1/instances/{instanceName}/restores/{id}/'
    },
    'list': {
      'methods': ['get'],
      'path': '/v1.1/instances/{instanceName}/restores/'
    },
    'restore': {
      'methods': ['post'],
      'path': '/v1.1/instances/{instanceName}/restores/'
    }
  }
});

var RestoreConstraints = {
  instanceName: {
    presence: true,
    length: {
      minimum: 5
    },
    id: {
      number: true
    }
  }
};

/**
 * OO wrapper around instance restores {@link # endpoint}.
 * @constructor
 * @type {Backup}

 * @property {Number} id
 * @property {String} instance
 * @property {String} created_at
 * @property {String} updated_at
 * @property {Number} size
 * @property {String} status
 * @property {String} status_info
 * @property {String} description
 * @property {String} label
 * @property {Object} [links = {}]
 */
var Restore = (0, _stampit2.default)().compose(_base.Model).setQuerySet(RestoreQuerySet).setMeta(RestoreMeta).setConstraints(RestoreConstraints).methods({
  restore: function restore() {
    var payload = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    var meta = this.getMeta();
    var path = meta.resolveEndpointPath('restore', this);

    return this.makeRequest('POST', path, { payload: payload });
  }
});

exports.default = Restore;
module.exports = exports['default'];
